### ISSUE TITLE
Client hangs with TRACY_NO_CALLSTACK without TRACY_NO_SYSTEM_TRACING

### ISSUE THREAD

**Issue by dmirys** [wolfpld/tracy#589]
I'm trying to reduce amount of data collected by tracy.
For that purpose I use the following set of flags:
* TRACY_ON_DEMAND
* TRACY_NO_CALLSTACK
* TRACY_NO_CODE_TRANSFER
* TRACY_NO_SAMPLING

My app hangs at the exit in the [infinite loop](https://github.com/wolfpld/tracy/blob/master/public/client/TracyProfiler.cpp#L1995-L2030), after sending terminate command to the server. Weanwhile server processed terminate instruction and detects there are [non zero m_pendingCallstackFrames](https://github.com/wolfpld/tracy/blob/master/server/TracyWorker.cpp#L2883-L2891). Looks like it supposed to get more data from client in such case. I think some misslogic here: client is waiting for data from server, while server is waiting data from client, while the client said to terminate.

With debugger I found that client sends [QueueType::CallstackSample commands]((https://github.com/wolfpld/tracy/blob/master/public/client/TracyProfiler.cpp#L2341-L2352)) to the server during tracy session. Disabling system tracing with `TRACY_NO_SYSTEM_TRACING` solves the issue.
Am I lost something useful with this option?
What is a correct way to solve the problem?
I'm ready to test some ideas.

**Comment by mo-tenstorrent** [comment]
@dmirys Did you find a solution here. I am also running into a similar issue. 

I hang at exit even with `TRACY_NO_SYSTEM_TRACING`

**Comment by dmirys** [comment]
No. Still using workaround. Here is the full set of flags that I use:
PUBLIC TRACY_NO_CALLSTACK
TRACY_NO_SYSTEM_TRACING
TRACY_NO_CODE_TRANSFER

Sorry, you have to debug if it doesn't help.

**Comment by wolfpld** [comment]
> It is specifically on the join in `Thread` destructor call in `~Profiler`.
> 
> https://github.com/wolfpld/tracy/blob/005d0929035bdc13f877da97c6631c0f2c98673e/public/client/TracyProfiler.cpp#L1564

What are other threads doing when this join is pending? Specifically, the `LaunchWorker` thread?

**Comment by mo-tenstorrent** [comment]
It is stuck on `poll`
<img width="878" alt="Screenshot 2024-05-24 at 1 39 35 PM" src="https://github.com/wolfpld/tracy/assets/109363418/c61a335c-cdbc-4e4a-ae87-587b5645f7d1">

**Comment by mo-tenstorrent** [comment]
Was wondering if there was any progress here? Is there any more data I can provide or tests to run?

**Comment by wolfpld** [comment]
According to the man page,

> If the value of timeout is 0, poll() shall return immediately.

The glibc code at frame #0 in your call stack is basically is doing a syscall into the kernel, which shouldn't be affected by the program environment.

I don't know what's wrong here, it shouldn't be happening.

**Comment by mo-tenstorrent** [comment]
Did some more digging on this, and actually the `poll()` function is behaving as expected. 

The issue is that when `TRACY_NO_CALLSTACK` is set, the last forever loop in `Profiler::Worker` doesn't break. 
<img width="600" alt="Screenshot 2024-07-17 at 11 33 02 AM" src="https://github.com/user-attachments/assets/f7f718ca-ad23-4bcb-b9b0-d6efae7b1867">

We keep Polling in the `HasData()` function and for some reason we are not receiving any data, let alone `ServerQueryDisconnect` to get out of the loop. I have confirmed this by  simple printfs. The loop keeps printing, so does HasData(). 

I tried the latest v0.10 commits to build the capture tool and our repo and the issue persists.

**Comment by JeffChorizo** [comment]
Chiming in to say that I encountered the same problem on shutdown while the server is connected, even though I am manually invoking RequestShutdown(), the manual shutdown bool is not respected by the sampling thread, unlike other worker threads.  In the very last for loop where the client wants to handle remaining server queries, what can happen is the last callstack decode request takes long enough for the next sample to happen, so the server requests another callstack decode, and the process repeats forever.

I worked around the issue by defining the environment variable TRACY_NO_SYS_TRACE=1 which turns off the sampling thread.  Might I suggest a better fix is to turn off the sampling thread when manual shutdown is true.

**Comment by adahlkvist-feral** [comment]
I have had a similar hang on disconnecting, in the sense that the client's worker thread hangs, not the desktop GUI. That means I can't reconnect.

Repro steps:
    * Turn on TRACY_ON_DEMAND
    * Use a platform with no defined TRACY_HAS_CALLSTACK (or undef it yourself to simulate this).
    * Connect to client
    * Profile for a bit, then press discard
    * Press "reconnect"
    * Observe that it never reconnects

The issue is here:

`while( m_symbolsBusy.load( std::memory_order_acquire ) ) { YieldThread(); }`

`m_symbolsBusy` will never be set to false since `Profiler::SymbolWorker` doesn't exist when TRACY_HAS_CALLSTACK is not defined.

I think the solution is to just guard that loop by `ifdef TRACY_HAS_CALLSTACK` :)

**Comment by wolfpld** [comment]
e59601cf1
